classes/engine.php

```diff
@@ -672,6 +680,85 @@ private function batch_add_documents($jsonpayload, $isdoc=false, $sendnow=false)
         return $numdocsignored;
     }
 
+    /**
+     * Parse the payload to extract document data with contextid.
+     *
+     * @return array Array of document data indexed by position.
+     */
+    private function parse_payload_documents(): array {
+        $documents = [];
+
+        if (empty($this->payload)) {
+            return $documents;
+        }
+
+        try {
+            $lines = explode("\n", trim($this->payload));
+
+            // Process pair of lines (metadata and document data).
+            $docindex = 0;
+            for ($i = 0; $i < count($lines); $i += 2) {
+                if (isset($lines[$i]) && isset($lines[$i + 1])) {
+                    $metadata = json_decode($lines[$i], true);
+                    $docdata = json_decode($lines[$i + 1], true);
+                    if ($metadata && $docdata) {
+                        $documents[$docindex] = [
+                            'metadata' => $metadata,
+                            'document' => $docdata,
+                            'id' => $docdata['id'] ?? null,
+                            'contextid' => $docdata['contextid'] ?? null,
+                            'areaid' => $docdata['areaid'] ?? null,
+                            'itemid' => $docdata['itemid'] ?? null,
+                            'modified' => $docdata['modified'] ?? null,
+                        ];
+
+                        $docindex++;
+                    }
+                }
+            }
+        } catch (\Exception $e) {
+            debugging('Error parsing payload documents: ' . $e->getMessage(), DEBUG_DEVELOPER);
```

Comment:
I'm a little concerned here. This method generally designed for getting some extra data for logging, but can actually fails a whole run because it can't parse $this->payload. In what situations parsing can cause an exception?
Can we please cover this method somehow with a good set of tests? Currently it's not very clear what would be expected as this->payload. And I assume tests could help to clarify that and confirm that this method works as expected.